commisery/commit.py

```diff
@@ -12,271 +12,396 @@
 # See the License for the specific language governing permissions and
 # limitations under the License.
 
-import os
+from collections import namedtuple
 import re
+import typing
 
-from dataclasses import dataclass, field
-from typing import Any, Optional, Sequence
-
-from commisery.config import Configuration
+import git
 
 BREAKING_CHANGE_TOKEN = "BREAKING CHANGE"
 
-CONVENTIONAL_COMMIT_REGEX = re.compile(
-    r"""
-        # 1. Commits MUST be prefixed with a type, which consists of a noun, feat, fix, etc.,
-        # followed by the OPTIONAL scope, OPTIONAL !, and REQUIRED terminal colon and space.
-        (?P<type>\w+)?
-        # 4. A scope MAY be provided after a type. A scope MUST consist of a noun describing a
-        # section of the codebase surrounded by parenthesis, e.g., fix(parser):
-        (?: \( (?P<scope> [^()]* ) \) )?
-        # 1. Commits MUST be prefixed with a type, which consists of a noun, feat, fix, etc.,
-        # followed by the OPTIONAL scope, OPTIONAL !, and REQUIRED terminal colon and space.
-        (?P<breaking_change>((\s*)+[!]+(\s*)?)?)
-        # 5. A description MUST immediately follow the colon and space after the type/scope prefix.
-        # The description is a short summary of the code changes, e.g., fix: array parsing issue when
-        # multiple spaces were contained in string.
-        (?P<separator>((\s+)?:?(\s+)?))
-        # 5. A description MUST immediately follow the colon and space after the type/scope prefix.
-        # The description is a short summary of the code changes, e.g., fix: array parsing issue when
-        # multiple spaces were contained in string.
-        (?P<description>.*)
-        """,
-    re.VERBOSE,
-)
-
-FOOTER_REGEX = re.compile(
-    r"""
-        # 8. One or more footers MAY be provided one blank line after the body. Each footer MUST consist
-        # of a word token, followed by either a :<space> or <space># separator, followed by a string value
-        # (this is inspired by the git trailer convention).
-        #
-        # 9. A footer’s token MUST use - in place of whitespace characters, e.g., Acked-by (this helps
-        # differentiate the footer section from a multi-paragraph body). An exception is made for
-        # BREAKING CHANGE, which MAY also be used as a token.
-        ^(?P<token>[\w\- ]+|BREAKING\sCHANGE)(?::[ ]|[ ](?=[#]))(?P<value>.*)
-    """,
-    re.VERBOSE,
-)
+COMMIT_TYPE = typing.Union["CommitMessage", "ConventionalCommit"]
```

Comment:
> Why don't you just define this without a forward reference by placing it after the definition of these types in this same module?

Personal preference. I like my constants to be defined at the top and will gladly pay the price of a forward reference for it.

> Also, because ConventionalCommit inherits from Commit you don't even need COMMIT_TYPE. I.e. assert isinstance(ConventionalCommit(), CommitMessage). So just use the base class for a type hint instead please...

I am aware. I've done this solely to be more explicit towards the human reading it.

Reply:
I'll switch to using the base class to move this forward.